### ISSUE TITLE
Allow variable expansion on *decorated* command statements

### ISSUE THREAD

**Issue by floam** [fish-shell/fish-shell#3595]
I think on the whole we are very happy not expanding variables in the command position in fish. But why does the parser not allow variables to expand when used as arguments for builtins like `command`? The rules are rather strange and handed on down from the parser with these.

## Presently
It's hard to understand why this works: `set -l prefix /usr/local/bin/; ls "$prefix/fish"`
But not:
```console
$ set -l prefix /usr/local/bin/; exec "$prefix/fish"
Variables may not be used as commands. In fish, please define a function or use 'eval "$path/fish"'.
```
It's inconsistent.

The "solution" is `eval`, but the `eval` function can be terribly difficult to use effectively or safely. We should never be suggesting this for mundane tasks. We have code like this, that only had the `string escape` thing added recently because nobody on the team noticed "Fish 2.3.0.app" in the data directory path will break it.

```fish
eval (string escape $__fish_datadir/tools/web_config/webconfig.py) $initial_tab
```
Because of hard-to-understand stuff happening with quoting, escaping, spaces, I doubt most people would happen upon a correct, good way to use eval without a lot of trial and error. 

`$fish_datadir/tools/web_config/webconfig.py`shouldn't be a valid command, but what if we allow this: `command "$fish_datadir/tools/web_config/webconfig.py"` ?

## Proof of concept
I've been using fish with a simple change to the parser for a while and I think it's a great improvement. I simply allow variable expansion on commands, only decorated ones, like `command foo`, `builtin foo`, `exec foo`, not undecorated naked commands such as "foo".
```diff
diff --git i/src/parse_execution.cpp w/src/parse_execution.cpp
index 1297a2a..d2275cf 100644
--- i/src/parse_execution.cpp
+++ w/src/parse_execution.cpp
@@ -827,9 +827,15 @@ parse_execution_result_t parse_execution_context_t::populate_plain_process(
     wcstring cmd;
     bool got_cmd = tree.command_for_plain_statement(statement, src, &cmd);
     assert(got_cmd);
-
+    bool expanded;
     // Expand it as a command. Return an error on failure.
-    bool expanded = expand_one(cmd, EXPAND_SKIP_CMDSUBST | EXPAND_SKIP_VARIABLES, NULL);
+    if (tree.decoration_for_plain_statement(statement) == parse_statement_decoration_none) {
+        // undecorated statement - we do not expand variables
+        expanded = expand_one(cmd, EXPAND_SKIP_CMDSUBST | EXPAND_SKIP_VARIABLES, NULL);
+    } else {
+        // decorated statement, e.g. "command $foo" - we expand variables
+        expanded = expand_one(cmd, EXPAND_SKIP_CMDSUBST, NULL);
+    }
```
It's likely wrong - I don't know what I'm doing in the parser. But it's enough to get an idea of what I think would be the desired behavior improvement for now.
```console
$ uname
Darwin
$ command $history[1]
Darwin
$ set foo ls
$ command $foo /
Applications			bin				net
Developer			cores				private
Library				dev				sbin
Network				etc				test
System				home				tmp
Users				installer.failurerequests	usr
Volumes				keybase				var
$ command $foo /cores
loginwindow-120-20160913T174108Z	loginwindow-120-20160913T174317Z	loginwindow-120-20160913T174334Z
$ $foo / # still not allowed
Variables may not be used as commands. In fish, please define a function or use 'eval $foo'.
```
Spaces:
```console
$ set foo "echo hi" # this isn't a valid command name
$ command $foo there
fish: Unknown command 'echo hi'
fish: command $foo there
```
versus eval:
```console
$ set __fish_bin_dir ~/Downloads/fish\ 2.app/Contents/Resources/base/bin/
$ eval "$__fish_bin_dir/fish" -v
fish: Unknown command '/Users/floam/Downloads/fish'
$ command "$__fish_bin_dir/fish" -v
fish, version 2.3.0
```

What do folks think?

**Comment by faho** [comment]
Personally, I tend to agree.

There are plenty of cases where you need to "run a variable" (think $PAGER), and `eval` is dangerous. I think I've said that before.

However, there are two things this would be missing:

- A way to explicitly run a function
- A way to run "something", whether it is a command, a builtin or a function

For this, we'd most likely need another keyword/builtin - `call` or so?

**Comment by krader1961** [comment]
Isn't this a duplicate of issue #154?

**Comment by floam** [comment]
> A way to explicitly run a function
That seems like a missing feature in fish, for sure, even without allowing variable expansion in these statements.

Also, one thing that currently doesn't work with my patch above is `set foo string; builtin $foo`, it only works for `command` and `exec` right now.

**Comment by floam** [comment]
> Also, one thing that currently doesn't work with my patch above is set foo string; builtin $foo, it only works for command and exec right now.

This was really vexing me because it looked like it should be working, but I found a spot where it was looking for a particular error a little earlier than I'd expect for decorated builtins. It just wasn't expanding the command at all. Seems to fix it:

```diff
diff --git i/src/parse_util.cpp w/src/parse_util.cpp
index 35ccbca..2a9fada 100644
--- i/src/parse_util.cpp
+++ w/src/parse_util.cpp
@@ -1287,10 +1287,12 @@ parser_test_error_bits_t parse_util_detect_errors(const wcstring &buff_src,
                     }
 
                     // Check that we don't do an invalid builtin (issue #1252).
-                    if (!errored && decoration == parse_statement_decoration_builtin &&
-                        !builtin_exists(command)) {
-                        errored = append_syntax_error(&parse_errors, node.source_start,
+                    if (!errored && decoration == parse_statement_decoration_builtin) {
+                        expand_one(command, EXPAND_SKIP_CMDSUBST, NULL);
+                        if (!builtin_exists(command)) {
+                            errored = append_syntax_error(&parse_errors, node.source_start,
                                                       UNKNOWN_BUILTIN_ERR_MSG, command.c_str());
+                        }
                     }
                 }
             }
```

**Comment by floam** [comment]
Probably no reason not to also expand command substitutions.

**Comment by floam** [comment]
I'm going to make this a PR since I have code, intended to fix #154 which I hadn't noticed before doing this, and close this as a duplicate. Discussion can happen there.

**Comment by floam** [comment]
Duplicate of #154 - see PR #3596.
